File: extension-system/sdk-python/mocu_extension_sdk/embedding.py
from __future__ import annotations

from typing import Any

from .protocol import JsonRpcProtocol


class EmbeddingApi:
    """
    Lets an extension create embedding vectors with Mocu's configured
    embedding model (whatever provider/API key the user set up in Mocu).
    """

    def __init__(self, protocol: JsonRpcProtocol) -> None:
        self._protocol = protocol

    def embed(self, texts: list[str]) -> dict[str, Any]:
        if not isinstance(texts, list) or not texts:
            raise ValueError(
                "Embedding embed requires a non-empty 'texts' array."
            )

        for index, text in enumerate(texts):
            if not isinstance(text, str) or not text.strip():
                raise ValueError(
                    f"Embedding texts[{index}] must be a non-empty string."
                )

        return self._protocol.request(
            "mocu.embedding.embed",
            {"texts": texts},
        )

    def embed_text(self, text: str) -> list[float]:
        """Convenience wrapper: embed a single text and return its vector."""
        result = self.embed([text])
        embeddings = result.get("embeddings")

        if not isinstance(embeddings, list) or not embeddings:
            raise ValueError(
                "Embedding provider did not return an embedding vector."
            )

        return embeddings[0]

File: extension-system/sdk-node/src/embedding.ts
import type {
  EmbeddingEmbedParams,
  EmbeddingEmbedResult,
} from "@mocu/extension-contracts";

import { EMBEDDING_EMBED_METHOD } from "@mocu/extension-contracts";

import { JsonRpcProtocolClient } from "./protocol-client.js";

/**
 * Lets an extension create embedding vectors with Mocu's configured
 * embedding model (whatever provider/API key the user set up in Mocu).
 */
export class EmbeddingApi {
  constructor(private readonly client: JsonRpcProtocolClient) {}

  async embed(
    params: EmbeddingEmbedParams,
  ): Promise<EmbeddingEmbedResult> {
    if (
      !params ||
      !Array.isArray(params.texts) ||
      params.texts.length === 0
    ) {
      throw new Error(
        "Embedding embed requires a non-empty 'texts' array.",
      );
    }

    const texts = params.texts.map((text, index) => {
      if (typeof text !== "string" || !text.trim()) {
        throw new Error(
          `Embedding texts[${index}] must be a non-empty string.`,
        );
      }

      return text;
    });

    return this.client.request<EmbeddingEmbedResult>(
      EMBEDDING_EMBED_METHOD,
      { texts },
    );
  }

  /**
   * Convenience wrapper: embed a single text and return its vector.
   */
  async embedText(text: string): Promise<number[]> {
    const { embeddings } = await this.embed({ texts: [text] });

    const embedding = embeddings[0];

    if (!Array.isArray(embedding)) {
      throw new Error(
        "Embedding provider did not return an embedding vector.",
      );
    }

    return embedding;
  }
}

export default EmbeddingApi;
